infrastructure/templates/public-api/deploy-stage-template.yml

```diff
@@ -0,0 +1,124 @@
+    strategy:
+      runOnce:
+        deploy:
+          steps:
+          - bash: echo "##vso[task.setvariable variable=dataProcessorFunctionAppName;]$(subscription)-ees-papi-fa-processor"
+            displayName: 'Set additional pipeline variables'
+          - download: EESBuildPipeline
+            displayName: 'Download Data Processor Function App ZIP file'
+            artifact: 'public-api-data-processor-$(upstreamPipelineBuildNumber)'
+          - checkout: self
+          - task: AzureCLI@2
+            displayName: 'Deploy bicep template to Azure'
+            inputs:
+              azureSubscription: ${{parameters.serviceConnection}}
+              scriptType: bash
+              scriptLocation: inlineScript
+              inlineScript: |
+                set -e
+                dataProcessorExists=`az functionapp list --resource-group $(resourceGroupName) --query "[?name=='$(dataProcessorFunctionAppName)']" | jq '. != []'`
+                
+                az deployment group create \
+                  --name 'DeployPublicApiInfrastructure$(upstreamPipelineBuildNumber)' \
+                  --resource-group $(resourceGroupName) \
+                  --template-file $(templateFile) \
+                  --parameters $(devParamFile) \
+                  --parameters \
+                      subscription='$(subscription)' \
+                      vNetName='$(vNetName)' \
+                      acrName='$(acrName)' \
+                      coreStorageAccountName='$(coreStorageAccountName)' \
+                      keyVaultName='$(keyVaultName)' \
+                      resourceTags='$(resourceTags)' \
+                      publicUrls='$(publicUrls)' \
+                      postgreSqlAdminName='$(postgreSqlAdminName)' \
+                      postgreSqlAdminPassword='$(postgreSqlAdminPassword)' \
+                      dockerImagesTag='$(upstreamPipelineBuildNumber)' \
+                      psqlDbUsersAdded='$(psqlDbUsersAdded)' \
+                      dataProcessorFunctionAppExists=$dataProcessorExists
+
+          - task: AzureCLI@2
+            displayName: Convert Bicep outputs to Azure Pipeline variables
+            inputs:
+              azureSubscription: ${{parameters.serviceConnection}}
+              scriptType: bash
+              scriptLocation: inlineScript
+              inlineScript: |
+                deploymentOutputs=$(az deployment group show \
+                --resource-group $(resourceGroupName) \
+                --name 'DeployPublicApiInfrastructure$(upstreamPipelineBuildNumber)' \
+                --query properties.outputs | jq -c 'to_entries[] | [.key, .value.value]')
+                
+                echo "$deploymentOutputs" | while IFS=$'\n' read -r c; do
+                  outputname=$(echo "$c" | jq -r '.[0]')
+                  outputvalue=$(echo "$c" | jq -r '.[1]')
+                  echo "##vso[task.setvariable variable=$outputname;]$outputvalue"
+                done
```

Comment:
I see there's quite a lot of pain here converting the output parameters to be part of the pipeline's environment! 😨

I really like the Azure Developer CLI (AZD) which can handle all this. It can be used in the pipeline in place of AZ. It wouldn't be easy to switch to using it now but maybe there's an opportunity for us to try using it later on.

It's stores of all the outputs after provisioning infrastructure to be used in the deployment task, and if deploying without provisioning infrastructure the environment variables can be initialised still using `azd env refresh` to resolve all the outputs for the deploy task.

Reply:
Ooh nice!  Yes, definitely one to explore later on.  I wish I could just stick at the infra work for longer as I see so many potential improvements but unfortunately that time doesn't exist right now!

Reply:
Luckily the chunk of code above was borrowed from another kindly person.  I can see it's pretty jarring though when taking in the pipeline as a whole, so I'm moving it out into a separate step template.